docs/design-approaches.md

```diff
@@ -0,0 +1,235 @@
+
+
+## Agent behavior
+
+Similarly to Runtime, there are different ways of running Agent inside the microVM.
+
+### One Agent per container
+
+There is a separate Agent instance per Create container request.
+
+**Cons**:
+* Someone should start new Agent on incoming Create container request.
+* Needs port negotiation. Each Agent reserves and listens to separate vsock port(s). Agent should let the Runtime know which port send requests to. Firecracker metadata service can be used to exchange this information, but it supports one-way communication only, from host to microVM.
+
+### One Agent per microVM
+
+Agent starts with microVM via startup script, lives as long as microVM is running, and serves fixed port Runtime knows about.
+
+**Pros**:
+* Agent always runs on well-known port, no need in port negotiation.
+
+**Cons**:
+* More logic in the agent to handle multiple containers and control information over the vsock
+
+
+## Multiplexing
+
+There are lots of data need to be transferred from Agent to Runtime (agents logs, for each container: stream container events/stdin/stdout/stderr). There are two major ways how this communication can be done:
+
+### Separate connection for each stream of data
+
+Agent creates several vsock channels and sends data separately (1 x service, ContainerCount x 3 stdin/out/err, 1 x logs).
+Runtime connects to Agent, reads data from socket channels and forwards to containerd/logfiles/etc.
+
+**Pros**:
+* Simplicity, no multiplexing needed.
+
+**Cons**:
+* Need to manage low-level socket connection, potentially handle retires, timeouts, encryption(?), etc.
+* Needs ports negotiation for each connection to let Runtime know which channel to listen to (and what kind of data to expect).
+* Final vsock design in `Firecracker` not clear ([issue to track](https://github.com/firecracker-microvm/firecracker/issues/650)). It can be either one `AF_UNIX` socket per guest or one `AF_UNIX` socket per `AF_VSOCK` port. In case of one socket per microVM it's not possible to send data via separate channels.
+
+### Use GRPC/TTRPC
+
+Rely on GRPC streaming [capabilities](https://grpc.io/docs/guides/concepts.html#server-streaming-rpc). In this case the only modification of proto interface is needed, GRPC generates everything else. Runtime knows how to stream data from Agent via strongly typed interface (for example `ReadStdout(containerID) returns (ReadStreamResponse)`).
+
+**Pros**:
+* GRPC/TTRPC can generate strongly typed interface for data exchange (see service example below)
+* No need in port negotiation, Agent's well-known port always used.
+
+**Cons**:
+* TTRPC doesn't support streaming (need to implement and submit a PR)
+* GRPC is heavier, potentially it consumes more resources ([need to clarify how much heavier it'll be for just one service](https://github.com/xibz/GRPCvsTTRPC/blob/master/README.md)).
+* Multiplexing (in general) may consume more memory & CPU
+
```

Comment:
There's a third option here, which is multiplexing at the connection layer instead of the application layer (i.e., underneath gRPC instead of within gRPC).  From our discussion with the Firecracker team, they are looking into options around Firecracker handling some aspects of the communication (i.e., exposing the correct vsock device inside the VM) and the outer application needing to handle other aspects (i.e., (de)multiplexing vsock datagrams from an AF_UNIX socket on the host.